Issue #13 Complete - View Purchase Details

katmdev opened the pull request:
## Description

<!-- What does this code change? Why did I choose this approach? Did I learn anything worth sharing? Reminder: This will be a publicly facing representation of your work (READ: help you land that sweet dev gig). -->

- Created new component to display item details (`Details.js`)
- Chose to create modal to display item details over new route to add some visual flair. Please let us know if you think this would be better rendered as a route in list, or if you think there is a better place to call the modal.
- Created new `details` state to hold item details. If there are details in state, triggered by the `details` button in list, the modal will show. Hitting `back` clears the details, causing the modal to hide.
- Employed an overlay to limit other interactions with the app while the modal is present and avoid errors.
- Created new item property of `addedDate` to track when item was initially added to a user's list and give predictions on next purchase date before the algorithm kicks in. Added `timestamp` variable to `firebase.js` in order to update `writeToFirestore` in `AddItem.js`
- Displayed item details. We chose to follow the wireframe by displaying "Next purchase date" over "Days to next purchase", but will revisit if needed.


## Related Issue

<!-- If you write "closes" followed by the Github issue number, it will automatically close the issue for you when the PR merges -->
Closes #13 

## Acceptance Criteria

<!-- Include AC from the Github issue -->

* From the shopping list, user can tap to navigate to a detailed view of the item
* Detail view includes the following information:
  * Name of the item
  * Last purchased date
  * Estimated number of days until the next purchase
  * Number of times the item has been purchased
* When in the “item details” view, a back arrow should appear in the header that takes the user back to the “list” view

## Type of Changes

<!-- Put an `✓` for the applicable box: -->

|     | Type                       |
| --- | -------------------------- |
|    | :bug: Bug fix              |
|  ✓ | :sparkles: New feature     |
|  ✓ | :hammer: Refactoring       |
|    | :100: Add tests            |
|    | :link: Update dependencies |
|    | :scroll: Docs              |

## Updates

### Before
<!-- If UI feature, take provide screenshots -->
![Screen Shot 2020-09-10 at 12 21 33](https://user-images.githubusercontent.com/58346557/92763168-45675a80-f361-11ea-8573-4c4da47a1a15.png)


### After
<!-- If UI feature, take provide screenshots -->
![Screen Shot 2020-09-10 at 12 22 17](https://user-images.githubusercontent.com/58346557/92763236-52844980-f361-11ea-82ac-58b4f5bb517d.png)
**Item with purchase history**
![Screen Shot 2020-09-10 at 12 28 08](https://user-images.githubusercontent.com/58346557/92763249-544e0d00-f361-11ea-88bc-d34a6af0d9d1.png)
**Item with no purchase history**
![Screen Shot 2020-09-10 at 12 28 32](https://user-images.githubusercontent.com/58346557/92763259-5748fd80-f361-11ea-8506-e284dea87c9c.png)


## Testing Steps / QA Criteria

<!-- Provide steps the other cohort members and mentors need to follow to properly test your additions. -->

- Be sure to create a new token when trying out this list, or use `filly aural verse`. Old tokens/items added before this feature was created will not have the `addedDate` field in Firestore and will show "NaN" for next purchase date if number of purchases is less than 2.

katmdev committed:
Created Details component, created modal in List

App.css updated styling to create background and overlay effect for modal. List: added button to show Details component, created ctate property for showDetails. Details: created button to go back or close details modal.

TianaJohnson committed:
Build details page

 Added build out information; details.name, working on purchase dates

TianaJohnson committed:
Date Display Formatting

TianaJohnson committed:
Arrow Formating

 used html unicode for arrow

 used flex box to better format the UI

katmdev committed:
Added addedDate to item object

Added additional timestamp to firestore/state for items to track date initially added to list, updated all associated functions, created timestamp variable in firestore.js, update styling of back button

katmdev committed:
removed unnecessary styling to details header

QMaximillian commented on src/App.css:
```diff
@@ -59,3 +60,28 @@ ul {
 .checkbox {
   margin: 10px;
 }
+.background {
+  position: absolute;
+  background-color: rgba(41, 43, 45, 0.75);
+  top: 0;
+  bottom: 0;
+  right: 0;
+  left: 0;
+  z-index: 10;
+  display: flex;
+  justify-content: center;
+  align-items: center;
+}
+.modal {
+  background-color: white;
+  border-radius: 10px;
+  padding: 20px;
+  position: relative;
+}
+
+.back-button {
+  position: absolute;
+  left: 15px;
+  top: 15px;
+  padding: 10px;
```
I'm wondering if it would make sense to switch the `px` to `rem` units to scale for users who change their font-sizes

katmdev replied:
My take away from that accessibility presentation was that we want font-related sizing to scale, but layout-related sizing to remain constant. I don't think we'd want the positioning or padding of the button to scale when increasing the font size so I chose px over rem.

laurenyz reviewed (approved):
Looks great!  Awesome job creating the modal!!

katmdev commented:
> Looks great! Awesome job creating the modal!!

I was inspired by the demo modals from https://react-bootstrap.github.io/components/modal/

QMaximillian commented on src/components/Details.js:
```diff
@@ -0,0 +1,63 @@
+import React from 'react';
+
+const Details = ({ details, setDetails }) => {
+  let numOfPurch = details.purchaseDates.length;
+  const lastPurch = numOfPurch
+    ? Math.max(...details.purchaseDates)
+    : details.addedDate;
+  let futurePurch = lastPurch + details.frequency;
```
I love how readable all of this code is! Everything is clearly laid out, and all variables names clearly clue anyone reading the code.

I just checked Safari, Chrome, and Firefox and each browser, or maybe Firestore, is guaranteeing insert order for the `purchaseDates` array. I think we can remove the call to `Math.max` because the `purchaseDates` array will always be sorted and we just need to grab the last element (the most recent purchase) in the `purchaseDates` array.

katmdev replied:
I used `Math.max` to keep things consistent with other components using the same value. Perhaps we should have just stashed last purchase date as a separate variable in Firebase to avoid this!

QMaximillian commented on src/components/List.js:
```diff
@@ -82,11 +85,15 @@ You cannot undo this action, and this item's purchase history will be lost.`,
               );
             })}
         </ul>
+        {details.name ? (
```
This absolutely works fine, but I noticed we have `lodash` in the project and it comes with an `isEmpty` method which might be more explicit than using `details.name`. I think it would be `import { isEmpty } from 'lodash` at the top, and then `!isEmpty(details)` in the ternary on line 94. I definitely think I'm being nitpicky because you both did awesome on this issue , but wanted to bring up another solution!

laurenmbeatty replied:
Yeah, I like this idea^. Not a deal-breaker, for sure, but it's nice to not just rely on the name.

Another way to do this would be 
`{details.name && (
     <Details details={details} setDetails={setDetails} />
   )}`

...no ternary.

katmdev replied:
I haven't used lodash but was looking at isEmpty earlier this week! I'll give it a look and implement.

laurenmbeatty commented on src/components/Details.js:
```diff
@@ -0,0 +1,63 @@
+  const lastPurch = numOfPurch
+    ? Math.max(...details.purchaseDates)
+    : details.addedDate;
+  let futurePurch = lastPurch + details.frequency;
```
It's nice to be consistent with `let` and `const`, when possible. In this case, they could be all `let` or all `const`, as nothing is being assigned, right?

laurenmbeatty commented on src/lib/firebase.js:
```diff
@@ -18,13 +18,14 @@ var firebaseConfig = {
 
 let fb = firebase.initializeApp(firebaseConfig);
 let db = firebase.firestore();
+let timestamp = firebase.firestore.Timestamp.now().toMillis();
```
Nice.

laurenmbeatty commented on src/lib/firebase.js:
```diff
@@ -18,13 +18,14 @@ var firebaseConfig = {
-  const currentPurchaseDate = firebase.firestore.Timestamp.now().toMillis();
+  const currentPurchaseDate = timestamp;
   if (existingPurchaseDates.length > 2) {
     const newPurchaseDates = [...existingPurchaseDates, currentPurchaseDate];
```
```suggestion
    const newPurchaseDates = [...existingPurchaseDates, timestamp];
```

laurenmbeatty commented on src/lib/firebase.js:
```diff
@@ -18,13 +18,14 @@ var firebaseConfig = {
   existingPurchaseDates,
 ) {
-  const currentPurchaseDate = firebase.firestore.Timestamp.now().toMillis();
+  const currentPurchaseDate = timestamp;
```
```suggestion
```
Could get rid of this, and just use `timestamp` directly.

laurenmbeatty commented on src/components/Details.js:
```diff
@@ -0,0 +1,63 @@
+            ? getTime(lastPurch)
+            : 'This item has no purchase history'}{' '}
+        </h4>
+        <h4>Next Purchase Date: {getTime(futurePurch)} </h4>
```
You could probably guard against showing `NaN` by doing something like
`Next Purchase Date: {futurePurch ? getTime(futurePurch) :  'N/A'}`
I didn't test that out...but could be something like that.

katmdev replied:
futurePurch contains frequency so it would return true but I like the idea and will see if I can implement something similar.

laurenmbeatty reviewed (approved):
I think it's fine to not navigate to the modal, or a detail page, even though the AC ask for it. One nice thing if you did the navigation, is you'd be able to share a link to a specific item with someone. 

Very nice work...I like the modal.

laurenmbeatty commented:
One thing you'd want to think about with the modal:

right now, when the modal opens, the focus doesn't shift to the modal, and someone using a keyboard won't be able to access the back button. If you use a library-created modal, this would most likely be taken care of for you. 

Strong recommendation for: https://reach.tech/dialog/

laurenmbeatty commented on src/components/Details.js:
```diff
@@ -0,0 +1,63 @@
+    <div className="background">
+      <div className="modal">
+        <div className="details-header">
+          <button className="back-button" onClick={() => setDetails({})}>
```
I would put the aria-label on the button itself.

laurenmbeatty commented:
Building an accessible modal:
https://medium.com/@islam.sayed8/trap-focus-inside-a-modal-aa5230326c1b

Another one:
https://assortment.io/posts/accessible-modal-component-react-portals-part-2

katmdev committed:
added recommendations from PR

updated aria lable for button in details, updated h4s to ps to make more consistent with other text, streamlined display of modal from List (will add lodash once looking into further), updated instances of timestamp in firebase to avoid repetition, will add more accessibility to modal in future once reviewing resources

katmdev committed:
resolved merge conflict

katmdev committed:
fixed button
